tests/v2/validation/upgrade/upgrade_test.go

```diff
@@ -0,0 +1,450 @@
+	testSession := session.NewSession(u.T())
+	u.session = testSession
+
+	client, err := rancher.NewClient("", testSession)
+	require.NoError(u.T(), err)
+
+	u.client = client
+
+	names := newNames()
+	u.names = names
+
+	clusterName := client.RancherConfig.ClusterName
+	require.NotEmptyf(u.T(), clusterName, "Cluster name to install resources is not set")
+
+	clusterID, err := clusters.GetClusterIDByName(client, clusterName)
+	require.NoError(u.T(), err)
+
+	projectName := names.core["projectName"]
+	project, err := projects.GetProjectByName(client, clusterID, projectName)
+	require.NoError(u.T(), err)
+
+	u.project = project
+
+	if project == nil {
+		projectConfig := &management.Project{
+			ClusterID: clusterID,
+			Name:      projectName,
+		}
+		createdProject, err := client.Management.Project.Create(projectConfig)
+		require.NoError(u.T(), err)
+		require.Equal(u.T(), createdProject.Name, projectName)
+		u.project = createdProject
+	}
+}
+
+func (u *UpgradeTestSuite) TestPreUpgrade() {
+	subSession := u.session.NewSession()
+	defer subSession.Cleanup()
+
+	client, err := u.client.WithSession(subSession)
+	require.NoError(u.T(), err)
+
+	u.T().Log("Creating a namespace")
+	createdNamespace, err := namespaces.CreateNamespace(client, u.names.random["namespaceName"], "{}", map[string]string{}, map[string]string{}, u.project)
+	require.NoError(u.T(), err)
+	assert.Equal(u.T(), createdNamespace.Name, u.names.random["namespaceName"])
+	u.namespace = createdNamespace
+
+	podTemplate := corev1.PodTemplateSpec{
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{
+				*testContainer,
+			},
+		},
+	}
+
+	u.T().Log("Creating a deployment with the test container")
+	createdDeployment, err := deployments.CreateDeployment(client, u.project.ClusterID, u.names.random["deploymentName"], u.namespace.Name, podTemplate)
+	require.NoError(u.T(), err)
+	assert.Equal(u.T(), createdDeployment.Name, u.names.random["deploymentName"])
+
+	u.T().Log("Waiting deployments to have expected number of available replicas")
+	err = charts.WatchAndWaitDeployments(client, u.project.ClusterID, u.namespace.Name, metav1.ListOptions{})
+	require.NoError(u.T(), err)
+
+	u.T().Log("Creating a daemonset with the test container")
+	createdDeamontSet, err := daemonsets.CreateDaemonSet(client, u.project.ClusterID, u.names.random["daemonsetName"], u.namespace.Name, podTemplate)
+	require.NoError(u.T(), err)
+	assert.Equal(u.T(), createdDeamontSet.Name, u.names.random["daemonsetName"])
+
+	secretTemplate := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: u.names.random["secretName"],
+		},
+		Data: map[string][]byte{
+			"test": []byte("test"),
+		},
+	}
+	u.T().Log("Creating a secret")
+	createdSecret, err := secrets.CreateSecret(client, secretTemplate, u.project.ClusterID, u.namespace.Name)
+	require.NoError(u.T(), err)
+	assert.Equal(u.T(), createdSecret.Name, u.names.random["secretName"])
+
+	podTemplateWithSecretVolume := corev1.PodTemplateSpec{
```

Comment:
In all these resource instantiations  Ugur, I think it would be best to write constructors for them, and put them in the respective extensions packages. Like the PodTemplateSpec, IngressSpec, etc. What do you think? These object setups are quite large, so I think was could benefit for having them written as constructors.

Reply:
Updated